From 28fa4c8bcab8df6db96f9a8f47199074a74aaca7 Mon Sep 17 00:00:00 2001
From: Anson Huang <Anson.Huang@nxp.com>
Date: Thu, 26 Jul 2018 21:32:03 +0800
Subject: [PATCH 4251/5242] MLK-19008 soc: imx: mu: do NOT clear RIEn for
 i.mx6sx and i.mx7d

commit  4a5aba390c6b6b09d90f0e2a572078c49acc82df from
https://source.codeaurora.org/external/imx/linux-imx.git

MU is shared between rpmsg driver and the multi-core
power management on i.MX6SX and i.MX7D, the RIE3 is
enabled by MU driver, but when rpmsg is probed, it
will call MU_Init and RIE3 will be clear and cause
multi-core power management never work, so do NOT
clear RIEn during MU initialization for i.MX6SX and
i.MX7D.

Signed-off-by: Anson Huang <Anson.Huang@nxp.com>
Reviewed-by: Peng Fan <peng.fan@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/soc/imx/mu/mx8_mu.c |   14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/drivers/soc/imx/mu/mx8_mu.c b/drivers/soc/imx/mu/mx8_mu.c
index 894c8fc..30c7fd8 100644
--- a/drivers/soc/imx/mu/mx8_mu.c
+++ b/drivers/soc/imx/mu/mx8_mu.c
@@ -8,6 +8,7 @@
 #include <linux/err.h>
 #include <linux/io.h>
 #include <linux/mx8_mu.h>
+#include <linux/of.h>
 
 static int version;
 
@@ -136,9 +137,18 @@ void MU_Init(void __iomem *base)
 			  ? MU_V10_ACR_OFFSET1 : MU_ACR_OFFSET1;
 
 	reg = readl_relaxed(base + offset);
-	/* Clear GIEn, RIEn, TIEn, GIRn and ABFn. */
-	reg &= ~(MU_CR_GIEn_MASK1 | MU_CR_RIEn_MASK1 | MU_CR_TIEn_MASK1
+	/* Clear GIEn, TIEn, GIRn and ABFn. */
+	reg &= ~(MU_CR_GIEn_MASK1 | MU_CR_TIEn_MASK1
 		 | MU_CR_GIRn_MASK1 | MU_CR_NMI_MASK1 | MU_CR_Fn_MASK1);
+
+	/*
+	 * i.MX6SX and i.MX7D have multi-core power management which need
+	 * to use RIE interrupts.
+	 */
+	if (!(of_machine_is_compatible("fsl,imx6sx") ||
+		of_machine_is_compatible("fsl,imx7d")))
+		reg &= ~MU_CR_RIEn_MASK1;
+
 	writel_relaxed(reg, base + offset);
 }
 
-- 
1.7.9.5

